main.py

```diff
@@ -53,24 +54,87 @@
 parser.add_argument('--multiple_scale',default='1', type=str,help='multiple_scale: e.g. 1 1,1.1  1,1.1,1.2')
 parser.add_argument('--test_dir',default='all',type=str, help='./test_data')
 
+parser.add_argument('--resume_epoch', default=0, type=int, help='resume from which epoch, if 0, no resume')
+parser.add_argument('--experiment_index', default=0, type=int, help='index of training time')
 # arguments for optimization
 parser.add_argument('--cdw', action='store_true', help='use cosine distance weight for model aggregation, default false' )
-parser.add_argument('--kd', action='store_true', help='apply knowledge distillation, default false' )
+parser.add_argument('--kd', action='store_true', help='apply knowledge distillation, default false')
+parser.add_argument('--kd_method', default='cluster', type=str, help='whole or cluster')
 parser.add_argument('--regularization', action='store_true', help='use regularization during distillation, default false' )
+parser.add_argument('--clustering', action='store_true', help='use clustering to aggregate models, fault false')
+parser.add_argument('--clustering_method', default='finch', type=str, help='method used for clustering, finch or kmeans')
+parser.add_argument('--max_distance', default=0.9, type=float, help='maximum distance in finch algorithm')
+parser.add_argument('--n_cluster', default=2, type=int, help='number of cluster in Kmeans')
+
+
+def save_checkpoint(server, clients, client_list, cpk_dir, epoch):
+    torch.save({
+        'epoch': epoch,
+        'server_state_dict': server.federated_model.state_dict(),
+        'client_list': [clients[c].cid for c in client_list],
+        'client_classifier': [clients[c].classifier.state_dict() for c in client_list],
+        'client_model': [clients[c].model.state_dict() for c in client_list]
+    }, os.path.join(cpk_dir, "{}.pth".format(epoch)))
+
+
+def load_checkpoint(path):
+    cpk = torch.load(path)
+    epoch = cpk['epoch']
+    server_state_dict = cpk['server_state_dict']
+    client_list = cpk['client_list']
+    client_classifier = cpk['client_classifier']
+    client_model = cpk['client_model']
+    return epoch, server_state_dict, client_list, client_classifier, client_model
+
 
 
 def train():
     args = parser.parse_args()
     print(args)
-    
+    if args.clustering:
+        clu = "clu"
+    else:
+        clu = "Nclu"
+
+    if args.cdw:
+        cdw = "cdw"
+    else:
+        cdw = "Ncdw"
+    if args.kd:
+        kd = "kd"
+    else:
+        kd = "Nkd"
+    if args.regularization:
+        reg = "reg"
+    else:
+        reg = "Nreg"
```

Comment:
Are these `if else` from line 95 - 110 only for setting the name for the checkpoint? It would be better to put them into a function.